client/securedrop_client/storage.py

```diff
@@ -66,6 +66,12 @@
 ).match
 
 
+class SDDatabaseError(Exception):
+    """
+    Handle SQLAlchemy exceptions and return relevant information to controller.
+    """
+
```

Comment:
@legoktm : I know we talked about not adding another Exception class, and using the existing ones. However, I feel like there's a case to be made here for a generic exception class that wraps SQLAlchemy exceptions, for 2 reasons:

1: far too many parts of the client are aware of sqlalchemy exceptions, including downloads.py, crypto.py, etc (see #2222); for errorhandling consistency and to avoid uncaught exceptions leading to app crashes, I would prefer db.py to anticipate all sqlalchemy errors and raise something db-agnostic to its callers; 2: I don't think we have a great answer in the code already (ie it's not a DownloadException etc).

For the purposes of this PR, I don't catch all SQLAlchemy exceptions and raise this one, but I'm proposing that for #2222. I also think it will be useful in that PR to be able to distinguish between "an 'anticipatable' sqlalchemyexception in our case (ie NoResultFound when searching for `one()` record would be anticipated under the sync/delete race condition) vs an unexpected SQLAlchemy exception.

I would welcome your thoughts though if you think there is a preferable way of doing this.

Reply:
I think the problem is that we're losing information here, we're turning a very specific `NoResultFound` error into a generic `SDDatabaseError`, which (in the future) could be *any* error, like db corruption or something else. And those types of different underlying errors should probably be handled differently; e.g. instead of having log messages like "likely deleted record", we should just know it absolutely is a deleted record and not something else.

IMO instead of using exceptions, I think we'd benefit from making the functions like:

```py3
def get_file(session: Session, uuid: str) -> File | None:
    try:
        return session.query(File).filter_by(uuid=uuid).one()
    except NoResultFound as e:
        return None
```

which is basically what you did in logic.get_file() :)

I think this addresses your concerns as the caller isn't aware of SQLAlchemy errors, it just knows there's a possibility that `get_file` won't return a file. The main advantage is now that it's encoded in the type system, mypy can now flag all the places this condition isn't being checked instead of us needing to manually remember to check the exception.

Reply:
This all makes sense, and I was probably chickening out a bit because returning `File | None`, `Message | None` etc would require a lot more client code changes, but is the better approach.

Zooming out a bit though, even if we don't do it in this case (since in this case we can create defined behaviour instead), I do still think it's worth encapsulating SQLAlchemy errors so that non db classes don't have to be aware of them; right now gui classes try-catch sqlalchemy errors and this feels messy and/or prone to problems. Are you open to that in a later PR, if the precondition is "make sure that we do the most with types that we can first"/make sure we don't lose error fidelity?

(The "probably a deletion" issue is going to be approximately true regardless of whether it's done at the return type level or the error level, because there could be other things that would lead to the missing entry other than a remote deletion, but you're right I don't have to be so cagey in the messages.)